### PR TITLE
Only one tree

### PR DESCRIPTION
**Pull request by emord**
No longer have to keep track of both the data and control tree.

The control tree is the master and we use the flags isControlOnly and isDataOnly whenever we need to discern between those.

### CODE REVIEW

File: tests/static/all_question_types.tsv [hunk 1]
```diff
@@ -28,6 +28,6 @@ Question	Type	Text (en)	Text (hin)	Audio (en)	Audio (hin)	Image (en)	Image (hin)
 /question22/question23/question28	Password	question28	question28									no
 /question22/question23/question29	Signature Capture	question29	question29									no
 /question22/question23/question7	Android App Callout	question7	question7									no
-	Base											no
 /question20	Hidden Value											no
-/question32	Hidden Value										1 + 2	no
\ No newline at end of file
+/question32	Hidden Value										1 + 2	no
+	Base											no
```
Comment (emord) [inline tests/static/all_question_types.tsv]: I changed this because it seemed to just fall out of the old implementation and I didn't see any comments about how we want this to stay the same.

Basically if something is only in the control path and at the end of the control body (like this piece of custom xml) it will end up getting put at the end.

Is this acceptable? I could only find this one edge case and it didn't seem to affect the end product (didn't change the xml on this same test)

File: src/form.js [hunk 2]
```diff
@@ -313,32 +307,13 @@ define([
             return writer.createXForm(this);
         },
         /**
-         * Goes through and grabs all of the data nodes (i.e. nodes that are
-         * only data nodes, possibly with a bind) without any kind of control.
-         * Returns a flat list of these nodes (list items are mugs).
-         */
-        getDataNodeList: function () {
```
Comment (emord) [inline src/form.js]: This was never used in vellum

File: src/form.js [hunk 3]
```diff
@@ -155,10 +155,6 @@ define([
 
         this.formName = 'New Form';
         this.mugMap = {};
-        this.dataTree = new Tree('data', 'data');
-        this.dataTree.on('change', function (e) {
-            _this.fireChange(e.mug);
-        });
         this.controlTree = new Tree('data', 'control');
```
Comment (millerdev) [inline src/form.js]: Can we rename this to `tree` now that there is only one?

File: src/form.js [hunk 4]
```diff
@@ -177,8 +173,7 @@ define([
 
     Form.prototype = {
         getBasePath: function () {
-            // the choice to use dataTree instead of controlTree is arbitrary
```
Comment (millerdev) [inline src/form.js]: Heh, when I wrote this comment I only meant it within the context of this function. Kind of amazing that you were able to run with it and figured out a way to apply it to all of Vellum.

File: src/form.js [hunk 5]
```diff
@@ -313,45 +307,19 @@ define([
...
                     return;
                 }
 
                 return node.getValue();
-            };
-
-            cList = this.controlTree.treeMap(treeFunc);
-            dList = this.dataTree.treeMap(treeFunc);
-
-            return util.mergeArray(cList, dList); //strip dupes and merge
```
Comment (millerdev) [inline src/form.js]: Looks like this was the only use of `mergeArray`, and  `_.uniq` does the same thing. Feel free to remove it from util.js.

File: src/form.js [hunk 6]
```diff
@@ -389,53 +357,28 @@ define([
...
-                        refMug = refMug.parentMug;
-                        position = (position === 'after' ? 'first' : 'last');
-                    }
-                }
-                this.controlTree.insertMug(newMug, position, refMug);
-            }
+            this.controlTree.insertMug(newMug, position, refMug);
         },
         /**
          * Move a mug from its current place (in both the Data and Control trees) to
```
Comment (millerdev) [inline src/form.js]: Should these documentation references to data and control trees be removed?

File: src/form.js [hunk 7]
```diff
@@ -446,23 +389,12 @@ define([
             this.fireChange(mug);
         },
         /**
-         * Walk merged data and control trees calling valueFunc for each mug
-         *
-         * If common nodes in merged tree branches are not in the same order
-         * then the order of nodes in the control tree will be preserved, and
-         * all nodes in the data tree that are not also in the control tree will
-         * be visited last.
```
Comment (millerdev) [inline src/form.js]: I assume this behavior has moved into the parser? (haven't gotten there yet) I think it's important that we preserve this behavior and explicitly mention it in the place where it is implemented.

File: src/form.js [hunk 8]
```diff
@@ -446,23 +389,12 @@ define([
...
          * single argument is the mug.
          */
-        mergedTreeMap: function (valueFunc) {
-            function getID(node) {
-                return node.getID() || node.getValue().ufid;
-            }
-            function hasControl(node) {
-                return !node.getValue().options.isDataOnly;
-            }
+        walkMugs: function (valueFunc) {
```
Comment (millerdev) [inline src/form.js]: This got quite a bit simpler, which is nice. It should be possible to make it simpler yet with `Tree.walk`. This example has not been tested, so may need to be tweaked:

```js
walkMugs: function (valueFunc) {
    function callback(mug, nodeID, processChildren) {
        if (mug !== null) {
            valueFunc(mug);
        }
        processChildren();
    }
    this.tree.walk(callback);
}
```

File: src/form.js (hunk 7 above)
Comment (millerdev) [inline src/form.js]: Got to the parser. It looks like the data tree order will be preserved, and the control tree order will be ignored/lost. I'm surprised that there are no tests that verify this comment. Please add a test that has data and control trees with mismatched order and verify that the control tree order is preserved after a parse/write cycle.

I'm not so worried about the last part of this comment "all nodes in the data tree that are not also in the control tree will be visited last." I think it's safe to omit that requirement if that's hard to implement.

File: src/writer.js [hunk 9]
```diff
@@ -156,7 +157,7 @@ define([
     };
 
     var createBindList = function (form, xmlWriter) {
-        form.dataTree.walk(function (mug, nodeID, processChildren) {
+        form.controlTree.walk(function (mug, nodeID, processChildren) {
             if(mug) {
```
Comment (millerdev) [inline src/writer.js]: This should probably say `&& !mug.isControlOnly`

File: tests/form.js [hunk 10]
```diff
@@ -111,40 +110,5 @@ define([
...
-        it("should prefer control tree order on merge", function () {
-            var form = new Form({}),
-                values = [];
-
-            form.dataTree = makeTree("data", ["a", "x1", "b", "x2", "c"]);
-            form.controlTree = makeTree("control", ["a", "c", "b"]);
-
-            form.mergedTreeMap(function (v) { values.push(v.id); });
-            assert.equal(values.join(" "), "a c b x1 x2");
-        });
```
Comment (millerdev) [inline tests/form.js]: Ha! I knew there were tests for this. These should be converted to XForm with equivalent structure and write tests to verify that the parser preserves order as specified here.

File: tests/static/all_question_types.tsv (hunk 1 above)
Comment (millerdev) [inline tests/static/all_question_types.tsv]: I'm not quite understanding why content from custom XML is showing up in this file, but what you're saying seems reasonable. I think we should probably make it skip this content entirely, but we can address that in a different PR.

Comment (emord) [inline tests/static/all_question_types.tsv]: I think it's because custom xml is turned into a ReadOnly Mug that can't be edited in Vellum, but still shows up. I'm not sure of it's use here, but didn't really investigate further

File: tests/form.js (hunk 10 above)
Comment (emord) [inline tests/form.js]: https://github.com/dimagi/Vellum/commit/03af95757856b736ead20fd0be2802a7fcafc2b8

File: src/form.js (hunk 8 above)
Comment (emord) [inline src/form.js]: Yep. that worked perfectly

File: tests/static/form/mismatch-tree-order.xml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        </input>
+        <group ref="/data/question2">
+            <label ref="jr:itext('question2-label')" />
+            <input ref="/data/question2/question3">
+                <label ref="jr:itext('question2/question3-label')" />
+            </input>
+        </group>
+        <input ref="/data/question4">
+            <label ref="jr:itext('question2/question4-label')" />
+        </input>
```
Comment (millerdev) [inline tests/static/form/mismatch-tree-order.xml]: The data tree order here is 1, 5, 2, 6, 4 with 5 and 6 being data-only, and control tree order is 1, 2, 4. Those are not mismatched trees, correct? I think you want to change this so the control tree order is 1, 4, 2.

The old test did this:
```js
form.dataTree = makeTree("data", ["a", "x1", "b", "x2", "c"]); // a b c
form.controlTree = makeTree("control", ["a", "c", "b"]);       // a c b
```

File: src/parser.js [hunk 12]
```diff
@@ -598,27 +598,55 @@ define([
...
+                if (numControlStructs > index) {
+                    break;
+                }
             }
-            if (mug.options.controlNodeChildren) {
+
+            form.tree.insertMug(mug, 'index', parentMug, realIndex);
+
+            if (mug.__className !== "ReadOnly" && mug.options.controlNodeChildren) {
+                var index1 = 0;
```
Comment (millerdev) [inline src/parser.js]: Very minor: might want to rename this to `child_index`. I would not even bring this up except it made code review a little harder. It wasn't easy to spot the difference between `index` and `index1`.

Comment (emord) [inline src/parser.js]: Variable names are hard

File: src/parser.js [hunk 13]
```diff
@@ -598,27 +598,55 @@ define([
...
+                    numControlStructs++;
+                }
+
+                if (numControlStructs > index) {
+                    break;
+                }
             }
-            if (mug.options.controlNodeChildren) {
+
+            form.tree.insertMug(mug, 'index', parentMug, realIndex);
```
Comment (millerdev) [inline src/parser.js]: Can this second insert be avoided (if `realIndex` is the same as the the mug's current index)? In most cases it will be unnecessary. It's a shame to have to add this algorithmic complexity to the parser, given that it's only aimed at rare forms with mismatched control and data trees.

Comment (emord) [inline src/parser.js]: I think I solved this in O(n) now

Comment (emord) [inline src/parser.js]: Err. maybe not O(n), but at least not as terrible as it once once.

File: tests/form.js [hunk 14]
```diff
@@ -112,39 +113,16 @@ define([
...
-        it("should merge data-only nodes with control nodes", function () {
-            var form = new Form({}),
-                values = [];
-
-            form.dataTree = makeTree("data", ["a", "x1", "b", "x2", "c"]);
-            form.controlTree = makeTree("control", ["a", "b", "c"]);
-
-            form.mergedTreeMap(function (v) { values.push(v.id); });
-            assert.equal(values.join(" "), "a x1 b x2 c");
-        });
```
Comment (millerdev) [inline tests/form.js]: After a bit more reflection, I think it would be good to rewrite this test as well. I'm not sure if there are any other tests that verify that the order of hidden values interspersed with other questions is preserved.

**Comment by millerdev** [comment]
Looks good. I'm tempted to be nitpicky about [names again](https://github.com/dimagi/Vellum/commit/317c7286026717552e3f564f040c3c1ba27511ae#diff-c1133de887698faf967f05bb89743782R133), but oh well... We can merge this once QA passes.
